cpp/src/arrow/acero/exec_plan.h

```diff
@@ -602,6 +603,9 @@ struct ARROW_ACERO_EXPORT QueryOptions {
   /// If this field is not set then it will be treated as kWarn unless overridden
   /// by the ACERO_ALIGNMENT_HANDLING environment variable
   std::optional<UnalignedBufferHandling> unaligned_buffer_handling;
+
+  /// \brief An optional stop-token for the query. Defaults to unstoppable.
+  StopToken stop_token = StopToken::Unstoppable();
```

Comment:
This API feels strange. Why would we want the user to pass a stop token in QueryOptions?

Reply:
What do you mean by user? If you mean end-user, then the end-user does not need to change this, because it is set by higher-level APIs. If you mean caller, then this options allows the caller to signal stopping via the stop-token. Specifically, [a stop is requested on the stop-source](https://github.com/apache/arrow/pull/35953/files#diff-70e843181eb977ed11ad131f79a0f54a0bf9e30a0cb752d440658ee0cc16cc34R962) from which [this stop-token was created](https://github.com/apache/arrow/pull/35953/files#diff-70e843181eb977ed11ad131f79a0f54a0bf9e30a0cb752d440658ee0cc16cc34R1005); then, [the stop-token is passed to the scheduler](https://github.com/apache/arrow/pull/35953/files#diff-70e843181eb977ed11ad131f79a0f54a0bf9e30a0cb752d440658ee0cc16cc34R194) which will detect the signal via the stop-token.

This approach was suggested by @westonpace, so he may be able to say more about why.

Reply:
> If you mean caller, then this options allows the caller to signal stopping via the stop-token. 
Why does the caller need to signal stopping via the stop-token? Doesn't https://github.com/apache/arrow/blob/main/cpp/src/arrow/acero/exec_plan.h#L107 already allow user to stop the execution?

Reply:
Hmm..OK. It feels strange that StopProducing does not work with Serial execution and we need a different way to stop an execution with serial execution. I would like to understand (1) why serial execution doesn't work with StopProducing now and (2) why we cannot make `StopProducing` to work (with serial execution.)

Reply:
I imagine it's possible to make `StopProducing` be sufficient to shut down the plan even if it is serially executing, based on the facilities in this PR. However, there may be unforeseen interactions so I'd suggest deferring this to a separate discussion/issue.

Reply:
@westonpace, the approach here is from an offline discussion with you that was active just a few days ago, and in my experiments `StopProducing` wasn't enough. What progress with `StopProducing` are you referring to? What is the approach you are currently in favor of? If it ends up being straightforward, then I'd be in favor too, but I'd like to first validate this is the case.

Reply:
If I recall the problem with `StopProducing` was that sometimes we were still seeing a hang even after `StopProducing` was called.  However, this was later revealed to be the "paused source node doesn't respect stop producing" problem.  Beyond that everything seemed to work correctly with `StopProducing`.

Reply:
Thanks @westonpace for chiming in. I would agree with you here.

Reply:
@rtpsw Should we revert this change?

Reply:
You're right.

Reply:
Done.